Implement the signal action

ekinanp opened the pull request:
Signal for core plugins will be implemented separately. The website will also be updated as part of that work.

Supports #548.

ekinanp committed:
Add the Signal action to the plugin package

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Implement Signal for external plugins

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Implement the fs/signal endpoint and the corresponding client call

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Add the signal command and supported signals to docs

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp commented:
I tested this by adding signal support to the `mock_docker` plugin. Here's what the output looks like for docs (note I used AWS plugin root's docs to represent a typical description):
```
wash . ❯ docs mock_docker/mock_containers_dir/chalice
This is the AWS plugin root. The AWS plugin reads the AWS_SHARED_CREDENTIALS_FILE
environment variable or $HOME/.aws/credentials and AWS_CONFIG_FILE environment
variable or $HOME/.aws/config to find profiles and configure the SDK. The profiles
it lists can be limited by adding

aws:
  profiles: [profile_1, profile_2]

to Wash’s config file.

The AWS plugin currently supports EC2 and S3. IAM roles are supported when configured
as described here. Note that currently region will also need to be specified with the
profile.

If using MFA, Wash will prompt for it on standard input. Credentials are valid for 1 hour.
They are cached under wash/aws-credentials in your user cache directory so they can be
re-used across server restarts. Wash may have to re-prompt for a new MFA token in response
to navigating the Wash environment to authorize a new session.

SUPPORTED SIGNALS
* stop
    Stop the container
* start
    Start the container
```

ekinanp commented:
As for testing signal, signal for the mock_docker plugin's containers writes the passed-in signal to a file. For example,
```
wash . ❯ signal foo mock_docker/mock_containers_dir/chalice
puppetlabs.wash/bad-action-request: Bad request for signal action on /Users/enis.inan/Library/Caches/wash/mnt568356764/mock_docker/mock_containers_dir/chalice: invalid signal foo. Valid signals are start, stop
wash . ❯ signal start mock_docker/mock_containers_dir/chalice
wash . ❯ exit
Goodbye!
➜  wash git:(signal) ✗ cat container_signal
start
```

ekinanp commented on api/types/signal.go:
```diff
@@ -0,0 +1,7 @@
+package apitypes
+
+// SignalBody encapsulates the payload for a call to a plugin's Signal function
+type SignalBody struct {
```
I chose this format because it's easier to extend it if `Signal` ever supports more arguments.

ekinanp commented on plugin/entrySchema.go:
```diff
@@ -66,13 +66,14 @@ func schema(e Entry) (*EntrySchema, error) {
+	Label               string            `json:"label"`
+	Description         string            `json:"description,omitempty"`
+	Singleton           bool              `json:"singleton"`
+	Signals             map[string]string `json:"signals,omitempty"`
```
I also considered supporting a tuple for methods like `(<method_name>, <method_schema>)` but that required a bit of a refactor and didn't generalize well to methods like `read`, `exec`, `schema` that didn't need `<method_schema>`. This way's also symmetric with wash ruby where there's a separate `signals` helper that lets you set the supported signals.

ekinanp replied:
Hmm, it looks like for Docker containers you can also send any one of the supported Linux signals (like `SIGHUP`, `SIGSTOP`, etc.) via `docker kill`. It'd be tedious for a user to have to specify these signals in the schema. Wonder if we could provide a short-hand for that.

Since these can be vendor-specific, I think we should change `Signals` to `map[signals]SignalSchema`

ekinanp replied:
Thinking about something like `map[string]SignalSchema` where `SignalSchema` is something like
```
type SignalSchema struct {
   Description
   Regex
}
```

If a Regex is provided, then we'll validate the provided signal using that. (At a higher-level, SignalSchemas w/ a Regex represent the schema of a group of signals, while w/o a Regex we are talking about a single Signal). Otherwise, we'll go w/ the signal name. This way, I can specify OS-specific signals as
```
"linux" => {
  "Description" => "Linux signals",
  "Regex" => "SIG*"
}
```

I'd prefer something like this instead of a hard-coded "linux" constant that people could use b/c that preserves Signal's general nature.

Could also change `Regex` to `Signals` to make validation easier.

ekinanp replied:
Yeah, I think storing this as an array to preserve the order makes sense.

Linux signals introduces the more general problem of "My API has a bunch of signals, many of which fit a specific group/category. I'd like to have a way of telling Wash that information without having to list every signal because not only is that tedious, but it will also annoy users of my plugin, especially since those users will (likely) be aware of what that specific group/category of signals represent when they view my entry's set of supported signals."

Having a Linux constant isn't a good solution to that problem, esp. when the general solution is (probably) not that much harder.

Not sure what you mean by extra flag though. What would that look like?

MikaelSmith replied:
Ah. I think I'd missed your last points. a Regex seems fine.

ekinanp committed:
Support signal groups

Signal groups are a useful way for plugin authors to compactly describe
a common set of signals, like e.g. the individual Linux signals.

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Use %w instead of %v in fmt.Errorf

%w incorporates Go 1.13's error wrapping

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp committed:
Fix linting error in methodWrappers signal test

Signed-off-by: Enis Inan <enis.inan@puppet.com>

ekinanp commented:
Ok, updated to support signal groups.
